File: src/clustr/tools/write/container_create.py
"""
Write tool for creating LXC containers.

Covers the common case: pull a template from storage, configure resources,
attach to the default bridge. Complex configs (multiple mounts, multiple
network interfaces, nesting) are better handled via the Proxmox UI.

destructiveHint = False: creating a container is additive.
readOnlyHint = False: mutating operation.
"""
from __future__ import annotations

import logging
from typing import Any

from mcp.types import TextContent, Tool

from clustr.proxmox.client import ProxmoxError, get_client

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Tool definition
# ---------------------------------------------------------------------------

TOOL_CREATE_CONTAINER = Tool(
    name="create_container",
    title="Create LXC Container",
    description=(
        "Create a new LXC container on a Proxmox node from a template. "
        "Requires a unique container ID, node name, hostname, OS template path, "
        "storage pool, and root disk size. "
        "The template must already exist on Proxmox storage — use the Proxmox "
        "UI or CLI to download templates first."
    ),
    inputSchema={
        "type": "object",
        "properties": {
            "node": {
                "type": "string",
                "description": "Node to create the container on (e.g. 'pve')",
            },
            "ctid": {
                "type": "integer",
                "description": "Unique container ID (100–999999). Must not already exist.",
                "minimum": 100,
                "maximum": 999999,
            },
            "hostname": {
                "type": "string",
                "description": "Container hostname (alphanumeric and hyphens, max 63 chars)",
            },
            "ostemplate": {
                "type": "string",
                "description": (
                    "Template path on Proxmox storage, e.g. "
                    "'local:vztmpl/debian-12-standard_12.7-1_amd64.tar.zst'"
                ),
            },
            "storage": {
                "type": "string",
                "description": "Storage pool for root disk (e.g. 'local-lvm'). "
                               "Use list_storage to find available pools.",
            },
            "disk_gb": {
                "type": "integer",
                "description": "Root disk size in GB (default: 8)",
                "minimum": 1,
                "default": 8,
            },
            "cores": {
                "type": "integer",
                "description": "Number of CPU cores (default: 1)",
                "minimum": 1,
                "maximum": 128,
                "default": 1,
            },
            "memory_mb": {
                "type": "integer",
                "description": "Memory in MB (default: 512)",
                "minimum": 128,
                "default": 512,
            },
            "swap_mb": {
                "type": "integer",
                "description": "Swap in MB (default: 512)",
                "minimum": 0,
                "default": 512,
            },
            "password": {
                "type": "string",
                "description": "Root password for the container. "
                               "If omitted, password login is disabled (SSH key recommended).",
            },
            "ssh_public_key": {
                "type": "string",
                "description": "SSH public key to inject into the container's authorized_keys.",
            },
            "unprivileged": {
                "type": "boolean",
                "description": "Create as unprivileged container (recommended, default: true)",
                "default": True,
            },
            "onboot": {
                "type": "boolean",
                "description": "Start container automatically on Proxmox boot (default: false)",
                "default": False,
            },
            "start_after_create": {
                "type": "boolean",
                "description": "Start the container immediately after creation (default: false)",
                "default": False,
            },
            "nameserver": {
                "type": "string",
                "description": "DNS nameserver IP (e.g. '1.1.1.1'). Optional.",
            },
        },
        "required": ["node", "ctid", "hostname", "ostemplate", "storage"],
    },
    annotations={
        "readOnlyHint": False,
        "destructiveHint": False,
    },
)


# ---------------------------------------------------------------------------
# Tool implementation
# ---------------------------------------------------------------------------

def _create_container(
    node: str,
    ctid: int,
    hostname: str,
    ostemplate: str,
    storage: str,
    disk_gb: int = 8,
    cores: int = 1,
    memory_mb: int = 512,
    swap_mb: int = 512,
    password: str = "",
    ssh_public_key: str = "",
    unprivileged: bool = True,
    onboot: bool = False,
    start_after_create: bool = False,
    nameserver: str = "",
) -> str:
    client = get_client()

    params: dict[str, Any] = {
        "vmid": ctid,
        "hostname": hostname,
        "ostemplate": ostemplate,
        "storage": storage,
        "rootfs": f"{storage}:{disk_gb}",
        "cores": cores,
        "memory": memory_mb,
        "swap": swap_mb,
        "unprivileged": 1 if unprivileged else 0,
        "onboot": 1 if onboot else 0,
        "net0": "name=eth0,bridge=vmbr0,ip=dhcp",
    }

    if password:
        params["password"] = password

    if ssh_public_key:
        params["ssh-public-keys"] = ssh_public_key

    if nameserver:
        params["nameserver"] = nameserver

    task_id = client.nodes(node).lxc.post(**params)

    if start_after_create:
        try:
            client.nodes(node).lxc(ctid).status.start.post()
        except Exception as exc:
            logger.warning("Container created but failed to start: %s", exc)

    return task_id


# ---------------------------------------------------------------------------
# Handler dispatcher
# ---------------------------------------------------------------------------

async def handle(tool_name: str, arguments: dict[str, Any]) -> list[TextContent]:
    if tool_name != "create_container":
        return [TextContent(type="text", text=f"Unknown tool: {tool_name}")]

    node = arguments.get("node", "").strip()
    ctid_raw = arguments.get("ctid")
    hostname = arguments.get("hostname", "").strip()
    ostemplate = arguments.get("ostemplate", "").strip()
    storage = arguments.get("storage", "").strip()

    if not node:
        return [TextContent(type="text", text="Error: 'node' is required.")]
    if not hostname:
        return [TextContent(type="text", text="Error: 'hostname' is required.")]
    if not ostemplate:
        return [TextContent(type="text", text="Error: 'ostemplate' is required.")]
    if not storage:
        return [TextContent(type="text", text="Error: 'storage' is required.")]
    if ctid_raw is None:
        return [TextContent(type="text", text="Error: 'ctid' is required.")]

    try:
        ctid = int(ctid_raw)
        disk_gb = int(arguments.get("disk_gb", 8))
        cores = int(arguments.get("cores", 1))
        memory_mb = int(arguments.get("memory_mb", 512))
        swap_mb = int(arguments.get("swap_mb", 512))
    except (TypeError, ValueError):
        return [TextContent(type="text", text="Error: ctid, disk_gb, cores, memory_mb, swap_mb must be integers.")]

    if ctid < 100:
        return [TextContent(type="text", text="Error: 'ctid' must be >= 100.")]

    password = arguments.get("password", "")
    ssh_public_key = arguments.get("ssh_public_key", "")
    unprivileged = bool(arguments.get("unprivileged", True))
    onboot = bool(arguments.get("onboot", False))
    start_after_create = bool(arguments.get("start_after_create", False))
    nameserver = arguments.get("nameserver", "").strip()

    try:
        task_id = _create_container(
            node=node,
            ctid=ctid,
            hostname=hostname,
            ostemplate=ostemplate,
            storage=storage,
            disk_gb=disk_gb,
            cores=cores,
            memory_mb=memory_mb,
            swap_mb=swap_mb,
            password=password,
            ssh_public_key=ssh_public_key,
            unprivileged=unprivileged,
            onboot=onboot,
            start_after_create=start_after_create,
            nameserver=nameserver,
        )
        return [
            TextContent(
                type="text",
                text=(
                    f"✅ Container **{hostname}** (ID: {ctid}) creation started on node **{node}**.\n"
                    f"Task ID: `{task_id}`\n\n"
                    f"**Config:**\n"
                    f"- CPU: {cores} core(s)\n"
                    f"- Memory: {memory_mb} MB\n"
                    f"- Swap: {swap_mb} MB\n"
                    f"- Root disk: {disk_gb} GB on `{storage}`\n"
                    f"- Template: `{ostemplate}`\n"
                    f"- Unprivileged: {'yes' if unprivileged else 'no'}\n"
                    f"- Start on boot: {'yes' if onboot else 'no'}\n\n"
                    f"Use `get_container_status` to check when the container is ready."
                ),
            )
        ]
    except ProxmoxError as exc:
        logger.error("Proxmox error in create_container: %s", exc)
        return [TextContent(type="text", text=f"Proxmox error: {exc}")]

File: clustr-ts/src/tools/write/containerCreate.ts
/**
 * Write tool for creating an LXC container from a template. confirm=false
 * (default) previews; confirm=true creates. A failed start_after_create is
 * surfaced rather than hidden.
 */

import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";

import { ProxmoxError, proxmoxPost } from "../../proxmox.js";
import { safe } from "../../safe.js";

interface CreateArgs {
  node: string;
  ctid: number;
  hostname: string;
  ostemplate: string;
  storage: string;
  disk_gb: number;
  cores: number;
  memory_mb: number;
  swap_mb: number;
  password: string;
  ssh_public_key: string;
  unprivileged: boolean;
  onboot: boolean;
  start_after_create: boolean;
  nameserver: string;
  bridge: string;
}

async function createContainer(
  a: CreateArgs,
): Promise<{ task: string; startStatus: string | null }> {
  const params: Record<string, string | number> = {
    vmid: a.ctid,
    hostname: a.hostname,
    ostemplate: a.ostemplate,
    storage: a.storage,
    rootfs: `${a.storage}:${a.disk_gb}`,
    cores: a.cores,
    memory: a.memory_mb,
    swap: a.swap_mb,
    unprivileged: a.unprivileged ? 1 : 0,
    onboot: a.onboot ? 1 : 0,
    net0: `name=eth0,bridge=${a.bridge},ip=dhcp`,
  };
  if (a.password) params.password = a.password;
  if (a.ssh_public_key) params["ssh-public-keys"] = a.ssh_public_key;
  if (a.nameserver) params.nameserver = a.nameserver;

  const task = String(await proxmoxPost(`/nodes/${a.node}/lxc`, params));

  let startStatus: string | null = null;
  if (a.start_after_create) {
    try {
      await proxmoxPost(`/nodes/${a.node}/lxc/${a.ctid}/status/start`);
      startStatus = "ok";
    } catch (err) {
      const msg = err instanceof ProxmoxError ? err.message : String(err);
      startStatus = `failed: ${msg}`;
    }
  }
  return { task, startStatus };
}

export function register(server: McpServer): void {
  server.registerTool(
    "create_container",
    {
      title: "Create LXC Container",
      description:
        "Create a new LXC container from a template. confirm=false (default) " +
        "returns the config that WOULD be created; call again with confirm=true " +
        "to create it. The template must already exist on Proxmox storage.",
      inputSchema: {
        node: z.string().describe("Node to create the container on"),
        ctid: z
          .number()
          .int()
          .min(100)
          .max(999999)
          .describe("Unique container ID (100–999999). Must not already exist."),
        hostname: z.string().describe("Container hostname (alphanumeric, hyphens)"),
        ostemplate: z
          .string()
          .describe(
            "Template path on storage, e.g. 'local:vztmpl/debian-12-standard_amd64.tar.zst'",
          ),
        storage: z
          .string()
          .describe("Storage pool for the root disk (use list_storage)"),
        disk_gb: z.number().int().min(1).default(8).describe("Root disk size in GB"),
        cores: z.number().int().min(1).max(128).default(1).describe("CPU cores"),
        memory_mb: z.number().int().min(128).default(512).describe("Memory in MB"),
        swap_mb: z.number().int().min(0).default(512).describe("Swap in MB"),
        password: z
          .string()
          .default("")
          .describe("Root password. If omitted, password login is disabled."),
        ssh_public_key: z
          .string()
          .default("")
          .describe("SSH public key to inject into authorized_keys"),
        unprivileged: z
          .boolean()
          .default(true)
          .describe("Create as unprivileged (recommended)"),
        onboot: z
          .boolean()
          .default(false)
          .describe("Start container on Proxmox boot"),
        start_after_create: z
          .boolean()
          .default(false)
          .describe("Start the container immediately after creation"),
        nameserver: z
          .string()
          .default("")
          .describe("DNS nameserver IP (optional)"),
        bridge: z.string().default("vmbr0").describe("Network bridge for eth0"),
        confirm: z
          .boolean()
          .default(false)
          .describe(
            "Must be true to actually create. When false (default), previews the config.",
          ),
      },
      annotations: { readOnlyHint: false, destructiveHint: false },
    },
    async (args) =>
      safe("create_container", async () => {
        const config =
          `**Config:**\n` +
          `- Node: \`${args.node}\`\n` +
          `- Container ID: \`${args.ctid}\`  Hostname: \`${args.hostname}\`\n` +
          `- CPU: ${args.cores} core(s)\n` +
          `- Memory: ${args.memory_mb} MB\n` +
          `- Swap: ${args.swap_mb} MB\n` +
          `- Root disk: ${args.disk_gb} GB on \`${args.storage}\`\n` +
          `- Template: \`${args.ostemplate}\`\n` +
          `- Network: \`eth0\` on bridge \`${args.bridge}\` (dhcp)\n` +
          `- Unprivileged: ${args.unprivileged ? "yes" : "no"}\n` +
          `- Start on boot: ${args.onboot ? "yes" : "no"}\n` +
          `- Start after create: ${args.start_after_create ? "yes" : "no"}\n`;

        if (!args.confirm) {
          return (
            `🔎 **Review: container not yet created.**\n\n${config}\n` +
            "Call `create_container` again with the same arguments plus `confirm=true` to create it."
          );
        }

        const { task, startStatus } = await createContainer({
          node: args.node,
          ctid: args.ctid,
          hostname: args.hostname,
          ostemplate: args.ostemplate,
          storage: args.storage,
          disk_gb: args.disk_gb,
          cores: args.cores,
          memory_mb: args.memory_mb,
          swap_mb: args.swap_mb,
          password: args.password,
          ssh_public_key: args.ssh_public_key,
          unprivileged: args.unprivileged,
          onboot: args.onboot,
          start_after_create: args.start_after_create,
          nameserver: args.nameserver.trim(),
          bridge: args.bridge,
        });

        let startLine = "";
        if (startStatus === "ok") {
          startLine = "▶️ Start requested. The container is booting.\n";
        } else if (startStatus) {
          startLine =
            `⚠️ The container was created but the start request failed: ${startStatus.slice(8)}. ` +
            "Wait for provisioning to finish, then call `start_container`.\n";
        }

        return (
          `✅ Container **${args.hostname}** (ID: ${args.ctid}) creation started on node **${args.node}**.\n` +
          `Task ID: \`${task}\`\n${startLine}\n${config}\n` +
          "Use `get_container_status` to check when the container is ready."
        );
      }),
  );
}
